### PR TITLE
Tbox output in ISO8601 format, Pandas & Numpy required, test changed

### LINKED ISSUE #78: Time extent in ISO 8601
Output of time extent (tbox) in the same date format (e.g., ISO 8601) for all supported files. Right now the output's format is different for each supported file. Relevant for issue #49  and  #75.

### LINKED ISSUE #81: (CSV) Time column names matches.
Time extraction only works if column name exactly matches "timestamp", "datetime", "time" or "date".

### LINKED ISSUE #82: (CSV) Enable time extraction of more than one column.
Geoextent only extracts one column with 'date' information. Some experiments record  start and end of a measurement, for this reason is relevant to take into account more than one column with 'date' information to extract the real time extent.

### PR DESCRIPTION
**Pull request by SbastianGarzon**
Closes #78
Closes #81 
Closes #82 

 **After** tbox extraction, the 2 dates (strings with an **unknown date format**) are parsed into a pandas timestamp with **the same datetime format** and then are transformed into [`ISO8601`](https://en.wikipedia.org/wiki/ISO_8601) date format (`'%Y-%m-%d'`).

To identify the 'common datetime format' the function [_guess_datetime_format_for_array](https://github.com/pandas-dev/pandas/blob/ca27ccae793e1281b67694a0c95501dc90771ba9/pandas/core/tools/datetimes.py#L72) is used to  extract the 'datetime' format for each string. In most of the cases, the ''guessed" formats are expected to be the same.  However for ambiguous dates (e.g 01.05.2010), the transformation takes into account the ''guessed" format of the other date. The transformation is successful only if the 'common datetime format' returns a valid `pandas.timestamp` date for both transformations. After the timestamps are transformed into  `pandas.Timestamp` both are finally parsed into `ISO8601`.

Expected geoextent test results are modified (to ISO8601 format).

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -3,3 +3,5 @@ geojson>=2.4.1
 pygeoj
 pyshp
 python-dateutil
+pandas
```
Comment (nuest) [inline requirements.txt:6]: Is it really required to add pandas? AFAICS that is a rather hefty tool if we only use date parsing..

Maybe you can document the different libraries you tried in the related issue?

File: geoextent/lib/extent.py [hunk 2]
```diff
@@ -99,7 +99,8 @@ def run(self):
             elif self.task == "tbox":
                 try:
                     if tbox:
-                        metadata["tbox"] = usedModule.getTemporalExtent(filePath)
+                        extract_tbox = usedModule.getTemporalExtent(filePath)
```
Comment (nuest) [inline geoextent/lib/extent.py:102]: Please check if you can improve the `getTemporalExtent()` function. Does it reliably work without flexible parsing methods? What about parsing to a proper data object already within the `get..` function and then in the next line only turning that date object into the desired string representation?

**Review by nuest — CHANGES_REQUESTED** [review]
I have a few questions.

Also, please take a look at https://github.blog/2013-01-22-closing-issues-via-commit-messages/ - it's very useful in cases where a specific commit closes a specific issue.

File: geoextent/lib/helpfunctions.py [hunk 3]
```diff
@@ -89,7 +93,37 @@ def validate(date_text):
...
+    else:
+        for i in time_form:
+            try:
+                time_1 = pd.to_datetime(list[0], format=i)
+                time_2 = pd.to_datetime(list[1], format=i)
+            except:
+                pass
+
+    if time_1 is not None and time_2 is not None:
+        time_1 = time_1.strftime('%Y-%m-%d')
```
[the remark this answers is not in the document]
Comment (SbastianGarzon) [reply geoextent/lib/helpfunctions.py]: Done ! Created as issue #80

File: requirements.txt (hunk 1 above)
Comment (SbastianGarzon) [reply requirements.txt:6]: As we require to parse all dates in the files in the same format, we need to handle ambiguous dates (e.g 01.05.2010). We can do this by extracting the preferred guessed format of all individual dates. After that, we have a list of a few potential 'date time formats' that we can use to parse all the dates. Some formats are only going to work for some dates (ambiguous dates) but not for other dates. This allow us to discard time formats and find the right one (the one format that correctly parse all dates). For this reason we required a function that returns the 'guessed' format. Pandas .datetimes is the only tool that returns information about the guessed format of string in the parsing process.

These feature is not available in other  'lighter tools' as dateparser: https://github.com/scrapinghub/dateparser/issues/341

File: geoextent/lib/extent.py (hunk 2 above)
Comment (SbastianGarzon) [reply geoextent/lib/extent.py:102]: I implemented this approach in the next commit !

File: geoextent/lib/helpfunctions.py [hunk 4]
```diff
@@ -98,32 +97,53 @@ def getDelimiter(csv_file):
...
     '''
 
-    time_1, time_2 = None, None
-    time_form = [time_format(np.array([list[0]])), time_format(np.array([list[1]]))]
-
-    if time_form[0] == time_form[1]:
-        time_1 = pd.to_datetime(list[0], format=time_form[0])
-        time_2 = pd.to_datetime(list[1], format=time_form[0])
+    date_time_format = None
+    num_sample = 30
```
Comment (nuest) [inline geoextent/lib/helpfunctions.py:108]: Please make this a function parameter with a default. If users use the Python API of the library, they might want to change `num_sample`. They should be able to change such a parameter as part of the main extraction function, too.

File: geoextent/lib/helpfunctions.py [hunk 5]
```diff
@@ -98,32 +97,53 @@ def getDelimiter(csv_file):
...
+    return date_time_format
 
-    if time_1 is not None and time_2 is not None:
-        time_1 = time_1.strftime('%Y-%m-%d')
-        time_2 = time_2.strftime('%Y-%m-%d')
 
-    bbox_iso8601 = [time_1, time_2]
+def date_parser(datetime_list):
+    '''
+    Function purpose: transform list of strings into ISO8601 ('%Y-%m-%d')
```
Comment (nuest) [inline geoextent/lib/helpfunctions.py:137]: Please update the user documentation to make clear that the temporal extent is only at "day" precision.

**Review by nuest — CHANGES_REQUESTED** [review]
Good improvement and bugfixing! Please update the docs to and expose the sampling configuration at least in the Python API (but not in the CLI).

File: geoextent/lib/helpfunctions.py (hunk 4 above)
Comment (SbastianGarzon) [reply geoextent/lib/helpfunctions.py:108]: Ok !

File: geoextent/lib/helpfunctions.py (hunk 5 above)
Comment (SbastianGarzon) [reply geoextent/lib/helpfunctions.py:137]: OK ! I'll update documentation for CLI and API adding the "day" precision.

**Review by nuest — CHANGES_REQUESTED** [review]
- add a test for changing the sample size via Python API
- user first and last, sample in the middle
- test files with some weird data formats
- add test that we do not get a result if there are mixed time formats in a file
- bump the minor version number in [`geoextent/__init.py`](https://github.com/o2r-project/geoextent/blob/master/geoextent/__init__.py)

File: geoextent/lib/extent.py [hunk 6]
```diff
@@ -99,38 +102,44 @@ def run(self):
             elif self.task == "tbox":
                 try:
                     if tbox:
-                        metadata["tbox"] = usedModule.getTemporalExtent(filePath)
+                        if usedModule.fileType == 'text/csv':
+                            extract_tbox = usedModule.getTemporalExtent(filePath, num_sample)
+                        else:
+                            if num_sample is not None:
+                                logger.warning("num_sample parameter NOT used, only required for csv")
```
Comment (nuest) [inline geoextent/lib/extent.py:109]: Is this really a "warning"?

File: geoextent/__init__.py [hunk 7]
```diff
@@ -1,3 +1,3 @@
 name = "geoextent"
 
-__version__ = '0.3.0'
\ No newline at end of file
+__version__ = '0.4.0'
```
Comment (nuest) [inline geoextent/__init__.py:3]: I changed this, before it was only the _bugfix_ portion that was increased.

File: geoextent/lib/helpfunctions.py [hunk 8]
```diff
@@ -66,30 +99,97 @@ def transformingArrayIntoWGS84(crs, pointArray):
...
\ No newline at end of file
+    return csv.reader(csv_file.readlines(), delimiter=dialect.delimiter)
+
+
+def get_time_format(time_list, num_sample):
+    '''
+    Function purpose: 'Guess' time format of a list of 'strings' by taking a representative sample
+    time_list:  list of strings \n
+    num_sample: size of the sample to determine time format \n
+    Output: time format in string format (e.g '%Y.%M.d')
```
Comment (nuest) [inline geoextent/lib/helpfunctions.py:136]: Is the output not always the same?

File: tests/test_api_csv.py [hunk 9]
```diff
@@ -33,96 +33,134 @@ def test_csv_extract_bbox_and_tbox_semicolon_delimiter():
...
+def test_csv_extract_tbox_random_sample_invalid():
+    result = geoextent.fromFile('tests/testdata/csv/3DCMTcatalog_TakemuraEPS.csv', bbox=False, tbox=True , num_sample = -1)
+    assert "bbox" not in result
+    assert "tbox" not in result
+
+def test_csv_extract_tbox_random_sample_value_larger_than_data():
+    result = geoextent.fromFile('tests/testdata/csv/3DCMTcatalog_TakemuraEPS.csv', bbox=False, tbox=True , num_sample = 1000000)
+    assert "bbox" not in result
+    assert "tbox" in result
+    assert result["tbox"] == ['2017-04-08', '2020-02-06']
```
Comment (nuest) [inline tests/test_api_csv.py:166]: Please document the test data source in https://github.com/o2r-project/geoextent/blob/master/tests/testdata/data-sources.md

File: geoextent/lib/extent.py (hunk 6 above)
Comment (nuest) [reply geoextent/lib/extent.py:109]: Please rephrase to "num_sample parameter is ignored, only applies to CSV"

**Review by nuest — APPROVED** [review]

File: geoextent/lib/handleCSV.py [hunk 10]
```diff
@@ -1,12 +1,15 @@
 import csv
-import os
 import logging
 from . import helpfunctions as hf
 
 fileType = "text/csv"
 
 logger = logging.getLogger("geoextent")
 
+search = { "longitude" : ["(.)*longitude","(.)*long(.)*", "^lon","lon$","(.)*lng(.)*", "^x","x$"],
```
Comment (nuest) [inline geoextent/lib/handleCSV.py:9]: Please add documentation for the supported options in the user docs!

Comment (SbastianGarzon) [reply geoextent/lib/handleCSV.py:9]: Ok ! I'm also including documentation regarding datetime format.

File: geoextent/lib/extent.py (hunk 6 above)
Comment (SbastianGarzon) [reply geoextent/lib/extent.py:109]: Raise warning if API user adds a `num_sample` parameter for a file that does not require it. The  `num_sample` parameter is only required for 'csv'. For other files this value is ignored.

**Pull request merged**